src/snowflake/connector/result_batch.py

```diff
@@ -666,9 +667,13 @@ def to_pandas(
         self, connection: SnowflakeConnection | None = None, **kwargs
     ) -> pandas.DataFrame:
         """Returns this batch as a pandas DataFrame"""
+        if self._dataframe is not None:
```

Comment:
I'm a little bit confused that why do we need to cache here? Doesn't the change in result_set.py already fix the issue?

Reply:
I'm not sure how expensive `to_pandas()` call is. The change in `result_set.py` leads to double calling `to_pandas()` when the dataframe is empty. The first call happens in https://github.com/snowflakedb/snowflake-connector-python/blob/0e9af4be39c63d728ecf89ac4003fb50dd098135/src/snowflake/connector/result_batch.py#L683.

Reply:
We can't cache results. This was considered when we did the distributed fetch redesign, but was rejected when we couldn't cache results without blowing up our memory footprint.

Consider the following:
1. I run a query with 37 gb of results.
2. I run `cursor.fetch_pandas_batches` to fetch results in batches so that I don't need to hold all results in memory at once.
3. This caching will force the user to hold all previously seen results in memory since [result sets hold references to all of their result batches](https://github.com/snowflakedb/snowflake-connector-python/blob/8108f63c053fdd74a739a1a0e8418856288f9753/src/snowflake/connector/result_set.py#L115) and now those would hold a reference to their previously seen results.

Reply:
I see. Then let's just do two `to_pandas()`. I think the overhead should be quite minimal given it is empty anyways.